### PR TITLE
Slurm Tutorial Page

### LINKED ISSUE #489: Add a Slurm tutorial
### What would you like to see added?

We don't have a tutorial about using Slurm!

Even our introduction assumes some knowledge of Cheaha that users might not know.

It would be beneficial to be able to point very new users at a tutorial for them to get a feel. Some topics to consider:

1. Hello world `sbatch` script.
2. Monitoring submissions with `squeue`
3. Checking past jobs with `salloc`
4. The most important/common flags (job-name, error, output | partition, time | cpus-per-task, mem, gres=gpu:#), their purpose and their effects.
5. Link to more advanced stuff and more detailed pages.

### PR DESCRIPTION
**Pull request by Premas**
# Pull Request

## Overview

Introducing a tutorial page for writing Slurm batch script for beginners, and to familiarize with distinct use cases.

## Proposed Changes
Added Slurm batch scripts for submitting sequential, and parallel jobs with concrete examples.

Example 1: Simple Batch job
Example 2: Sequential job
Example 3: Multijob using srun
Example 4: Array Job
Example 5: Multithread job
Example 6: GPU job
Example 7: Multinode job

## Related Issues
Fixes #489

### CODE REVIEW

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,528 @@
+# Writing Slurm Batch Jobs
+
+This Slurm tutorial serves as a hands-on guide for users to create Slurm batch scripts based on their specific software needs and apply them for their respective usecases.  It covers basic examples for beginners and advanced ones, including sequential and parallel jobs, array jobs, multithreaded jobs, GPU utilization jobs, and MPI (Message Passing Interface) jobs.
+
+## Strucutre of a Slurm Batch Job
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:5]: Typo: make it "Structure"

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+#!/bin/bash
+
+### Declaring Slurm configuration Options and specifying required resources
+#SBATCH --job-name=hostname ### Name of the job
+#SBATCH --nodes=1           ### Number of Nodes
+#SBATCH --ntasks=1          ### Number of Tasks
+#SBATCH --cpus-per-task=1   ### Number of Tasks per CPU
+#SBATCH --mem=1G            ### Memory required
+#SBATCH --partition=express ### Cheaha Partition
+#SBATCH --time=00:10:00     ### Estimated Time of Completion
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:41]: `### Estimated Time of Completion, 10 minutes`

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+
+```py linenums="1"
+#!/bin/bash
+
+### Declaring Slurm configuration Options and specifying required resources
+#SBATCH --job-name=hostname ### Name of the job
+#SBATCH --nodes=1           ### Number of Nodes
+#SBATCH --ntasks=1          ### Number of Tasks
+#SBATCH --cpus-per-task=1   ### Number of Tasks per CPU
+#SBATCH --mem=1G            ### Memory required
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:39]: `### Memory required, 1 gigabyte`

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+       JobID    JobName  Partition    Account  AllocCPUS      State ExitCode 
+------------ ---------- ---------- ---------- ---------- ---------- -------- 
+26127143          numpy    express      USER          1  COMPLETED      0:0 
+26127143.ba+      batch                 USER          1  COMPLETED      0:0 
+26127143.ex+     extern                 USER          1  COMPLETED      0:0
+```
+
+## Example 3: Parallel Jobs
+
+Multiple jobs or tasks can be executed simultaneously using `srun` within a single batch script. In this example, the same executable `python_script_new.py` is run in parallel with distinct inputs (line 17-19). The `&` symbol at the end of each line run these commands in background. The `wait` command (line 20) performs synchronization and ensures that all background processes and parallel tasks are completed before finishing. In Line 4, three tasks are requested as there are three executables to be run in parallel. Copy the batch script into a file named `multijob.job`. Use the `conda` environemnt `pytools-env`, as shown in [example2](../slurm/slurm_tutorial.md/#example-2-sequential-job).
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:141]: Add a sentence describing what happens with hardware. Overall script gets 3 CPUs, each `srun` gets 1 CPU.

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+### Perform addition on the array elemnts using numpy's sum function
+sum_result = np.sum(input_array)
+
+### Print Input Range and Sum
+print("Input Range: {} to {}, Sum: {}".format(start, end, sum_result))
+```
+
+The below ouptut shows that each line corresponds to the output of one parallel execution of python script with specific input ranges.
+
+```bash
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:193]: Please discuss the non-intuitive "out of order" results. Execution completion doesn't happen in the same order the jobs are submitted.

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+
+To test this example, copy the below script in a file named `hostname.job`. This job executes the `hostname` command (line 15) on a single node, using one task, one CPU core, 1 gigabyte of memory, with a time limit of 10 minutes. The output and error logs are directed to separate files with names based on their job name and ID (line 11 and 12). For a more detailed understanding of the individual parameters used in this script, please refer to the section on [Simple Batch Job](../slurm/submitting_jobs.md/#a-simple-batch-job). The following script includes comments, marked with `###`, describing their functions. We will utilize this same notating for annotating comments in subsequent examples.
+
+```py linenums="1"
+#!/bin/bash
+
+### Declaring Slurm configuration Options and specifying required resources
+#SBATCH --job-name=hostname ### Name of the job
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:35]: Space out the comments a bit further to the right. You can hold alt and click on additional lines (or use alt+up, alt+down) to add cursors. See here:

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+------------ ---------- ---------- ---------- ---------- ---------- -------- 
+27099591       multijob    express      USER          3  COMPLETED      0:0 
+27099591.ba+      batch                 USER          3  COMPLETED      0:0 
+27099591.ex+     extern                 USER          3  COMPLETED      0:0 
+27099591.0       python                 USER          1  COMPLETED      0:0 
+27099591.1       python                 USER          1  COMPLETED      0:0 
+27099591.2       python                 USER          1  COMPLETED      0:0 
+```
+
+## Example 4: Array Job
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:216]: I think it would be helpful to have one more example:

`## Example : Array Job - Files`

Please add a separate issue for this suggestion.

**Review by wwarriner — CHANGES_REQUESTED** [review]
Great work! I have some minor requests and one new suggestion that requires its own issue, which will add to this page.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 1 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md]: Fixed the typo

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 2 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md]: Incorporated the changes

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 3 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md]: Incorporated the changes

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 5 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md:193]: Discussed the result being out-of-order and possible approach to display the results in order.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 6 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md]: Incorporated the suggested changes.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 7 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md:216]: Created an issue for now.

**Comment by Premas** [comment]
@wwarriner Thanks for reviewing and suggesting changes. I have incorporated all the suggested changes.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 5 above)
Comment (wwarriner) [reply docs/cheaha/slurm/slurm_tutorial.md:193]: Looks great, thanks!

Please change "in out-of-order" to "out of order".

I would also change "If you want the results to be in order, ..." to something like "If the results must be in order to be correct, you will need to modify...". As you know, forcing an order gets rid of the parallelism. Either they're in order and not parallel, or out of order and parallel.

Of course, it may be possible to sort the results, if someone wants both :)

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 7 above)
Comment (wwarriner) [reply docs/cheaha/slurm/slurm_tutorial.md:216]: Thanks!

**Review by wwarriner — CHANGES_REQUESTED** [review]
Great!

Some small changes to the new paragraph about out of order execution and we can merge.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 5 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md:193]: Sure. Incorporated the changes.

File: mkdocs.yml [hunk 8]
```diff
@@ -53,7 +54,7 @@ plugins:
         account_management/xias_guest.md: account_management/xias/guest_instructions.md
         account_management/xias_sites.md: account_management/xias/pi_site_management.md
         account_management/xias_users.md: account_management/xias/pi_guest_management.md
-        cheaha/conda.md: cheaha/software/software.md#anaconda-on-cheaha
+        cheaha/conda.md: cheaha/software/software.md/#anaconda-on-cheaha
```
Comment (wwarriner) [inline mkdocs.yml:57]: Please revert this change. If I suggested it that was my mistake. It works the way it was before, we were entering the url incorrectly when we tested.

File: mkdocs.yml [hunk 9]
```diff
@@ -128,6 +129,7 @@ nav:
           - Practical Batch Array Jobs: cheaha/slurm/practical_sbatch.md
           - Managing Jobs: cheaha/slurm/job_management.md
           - GPUs: cheaha/slurm/gpu.md
+          - Slurm Tutorial: cheaha/slurm/slurm_tutorial.md
```
Comment (wwarriner) [inline mkdocs.yml:132]: Move this up to be between "SLURM/Introduction" and "SLURM/Submitting Jobs". Let's give it a little more visibility :)

File: docs/cheaha/slurm/slurm_tutorial.md [hunk 10]
```diff
@@ -0,0 +1,528 @@
+# Writing Slurm Batch Jobs
```
Comment (wwarriner) [inline docs/cheaha/slurm/slurm_tutorial.md:1]: I don't want to hold this one up any longer, so this doesn't have to be done as part of this review, it's a suggestion for the future.

I want to add a part at the beginning of the tutorial that tells people when/why they would want to use each type, so let's create an issue for it.

Something like:

```
## How do I know which one I need?

- Use 1 if...
- Use 2 if...
```

**Review by wwarriner — CHANGES_REQUESTED** [review]
Looks good, this one is ready to go after a couple minor changes. Let's create an issue to follow up and add a "How do I know which one I need?" section to give the reasoning for each.

Thank you @Premas!

File: mkdocs.yml (hunk 8 above)
Comment (Premas) [reply mkdocs.yml:57]: Fixed !

File: mkdocs.yml (hunk 9 above)
Comment (Premas) [reply mkdocs.yml:132]: Sure. Done.

File: docs/cheaha/slurm/slurm_tutorial.md (hunk 10 above)
Comment (Premas) [reply docs/cheaha/slurm/slurm_tutorial.md:1]: Sure. I will submit an issue.

**Review by wwarriner — APPROVED** [review]
